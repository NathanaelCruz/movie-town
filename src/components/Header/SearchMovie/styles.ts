import styled from 'styled-components'

export const SearchContent = styled.div`
  grid-area: busca;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  max-height: 100%;
  position: relative;
`
export const SearchInput = styled.input`
  width: 95%;
  height: 3rem;
  padding: 1rem 4rem 1rem 1rem;
  border-radius: 15px 0 0 15px;
  border: none;
  outline: none;

  &:active,
  &:focus {
    box-shadow: 0 0 5px 2px ${(props) => props.theme.colors.primary};
  }
`
export const SearchButton = styled.button`
  background-color: ${(props) => props.theme.colors.text};
  position: absolute;
  height: 4rem;
  width: 4rem;
  z-index: 10;
  right: 0;
  top: calc(50% - 2rem);
  border-radius: 50%;
  border: none;
  transition: 0.2s ease-in-out;

  &:hover {
    cursor: pointer;
    background-color: ${(props) => props.theme.colors.primary};
  }
`
