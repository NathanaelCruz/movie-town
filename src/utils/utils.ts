export const HostDataBase = 'http://localhost:3001'
