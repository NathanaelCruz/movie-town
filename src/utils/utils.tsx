import { ReactNode } from 'react'

import { ThemeProvider } from 'styled-components'
import theme from '../styles/theme'

export const HostDataBase = 'http://localhost:3001'

export function RenderWithTemplate(ComponentRender: ReactNode) {
  return <ThemeProvider theme={theme}>{ComponentRender}</ThemeProvider>
}
