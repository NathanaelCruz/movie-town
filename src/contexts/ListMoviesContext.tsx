import axios from 'axios'
import Modal from '../components/modal'
import { createContext, useState, ReactNode, useEffect } from 'react'
import { HostDataBase } from '../utils/utils'

type ListMoviesProviderProps = {
  children: ReactNode
}

type GenresProps = {
  genre: string
}

type ActorsProps = {
  name: string
  link: string
}

export type MovieProps = {
  id: string
  name: string
  slug: string
  folder: string
  rating: number
  synopsis: string
  year: number
  duration: number
  director: string
  genres: Array<GenresProps>
  actors: Array<ActorsProps>
}

type ListMoviesContextProps = {
  closeAndOpenModal: () => void
  listMovies: Array<MovieProps>
  listMoviesSearch: (str: string) => void
  setListMoviesWithAll: () => void
  searchForWordKey: () => void
  movieIsFinded: boolean
}

export const ListMoviesContext = createContext({} as ListMoviesContextProps)

export function ListMoviesProvider({ children }: ListMoviesProviderProps) {
  const [isOpen, setIsOpen] = useState(false)
  const [listMovies, setListMovies] = useState([])
  const [wordSearchCurrency, setWordSearchCurrency] = useState('')
  const [movieIsFinded, setMovieIsFinded] = useState(false)

  useEffect(() => {
    setListMoviesWithAll()
  }, [])

  function closeAndOpenModal() {
    setIsOpen(!isOpen)
  }

  function listMoviesSearch(wordKey: string) {
    setWordSearchCurrency(wordKey)
    axios.get(`${HostDataBase}/movies?q=${wordSearchCurrency}`).then((res) => {
      if (res.data.length > 0) {
        setListMovies(res.data)
        setMovieIsFinded(false)
      } else {
        setMovieIsFinded(true)
      }
    })
  }

  function searchForWordKey() {
    console.log(wordSearchCurrency)
    if (wordSearchCurrency === '') {
      closeAndOpenModal()
    } else {
      axios
        .get(`${HostDataBase}/movies?q=${wordSearchCurrency}`)
        .then((res) => {
          if (res.data.length > 0) {
            setListMovies(res.data)
            setMovieIsFinded(false)
          } else {
            setMovieIsFinded(true)
          }
        })
    }
  }

  function setListMoviesWithAll() {
    axios.get(`${HostDataBase}/movies`).then((res) => {
      setListMovies(res.data)
    })
  }

  return (
    <ListMoviesContext.Provider
      value={{
        closeAndOpenModal,
        listMovies,
        movieIsFinded,
        listMoviesSearch,
        setListMoviesWithAll,
        searchForWordKey
      }}
    >
      {children}
      {isOpen && <Modal />}
    </ListMoviesContext.Provider>
  )
}
