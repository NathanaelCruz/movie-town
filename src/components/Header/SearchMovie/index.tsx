import * as S from './styles'
import { SavedSearch } from '@styled-icons/material'
import { KeyboardEvent, useContext } from 'react'
import { ListMoviesContext } from '../../../contexts/ListMoviesContext'

const SearchMovie: React.FC = () => {
  const { listMoviesSearch, searchForWordKey } = useContext(ListMoviesContext)

  function searchWordKey(e: KeyboardEvent) {
    const valueCurrency = e.target as HTMLInputElement
    listMoviesSearch(valueCurrency.value)
  }

  return (
    <S.SearchContent>
      <S.SearchInput
        type="search"
        placeholder="Qual seu filme favorito?"
        onKeyUp={(event) => searchWordKey(event)}
      ></S.SearchInput>
      <S.SearchButton onClick={() => searchForWordKey()}>
        <SavedSearch size={24} color="#ffffff"></SavedSearch>
      </S.SearchButton>
    </S.SearchContent>
  )
}

export default SearchMovie
