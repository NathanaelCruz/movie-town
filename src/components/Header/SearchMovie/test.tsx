import { fireEvent, render, screen } from '@testing-library/react'
import { RenderWithTemplate } from '../../../utils/utils'
import { ListMoviesProvider } from '../../../contexts/ListMoviesContext'
import SearchMovie from './index'

describe('Barra de Busca', () => {
  it('Deverá renderizar um input', () => {
    const { container } = render(RenderWithTemplate(<SearchMovie />))

    expect(screen.getByRole('searchbox', {})).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Deverá aparecer o modal se não tiver nada no input', () => {
    render(
      RenderWithTemplate(
        <ListMoviesProvider>
          <SearchMovie />
        </ListMoviesProvider>
      )
    )

    fireEvent.click(screen.getByRole('button', {}))

    expect(screen.getByRole('dialog', {})).toBeInTheDocument()
  })
})
