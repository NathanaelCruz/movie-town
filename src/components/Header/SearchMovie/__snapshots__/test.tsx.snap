// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Barra de Busca Dever√° renderizar um input 1`] = `
.c3 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  grid-area: busca;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  max-height: 100%;
  position: relative;
}

.c1 {
  width: 95%;
  height: 3rem;
  padding: 1rem 4rem 1rem 1rem;
  border-radius: 15px 0 0 15px;
  border: none;
  outline: none;
}

.c1:active,
.c1:focus {
  box-shadow: 0 0 5px 2px #E74C3C;
}

.c2 {
  background-color: #2f3542;
  position: absolute;
  height: 4rem;
  width: 4rem;
  z-index: 10;
  right: 0;
  top: calc(50% - 2rem);
  border-radius: 50%;
  border: none;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.c2:hover {
  cursor: pointer;
  background-color: #E74C3C;
}

<div
  class="c0"
>
  <input
    class="c1"
    placeholder="Qual seu filme favorito?"
    type="search"
  />
  <button
    class="c2"
  >
    <svg
      aria-hidden="true"
      class="c3"
      color="#ffffff"
      fill="currentColor"
      focusable="false"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14zm-2.17-1.5l2.14-1.53 2.14 1.53-.83-2.46 2.15-1.5h-2.62L9.47 6l-.84 2.54H6l2.14 1.49z"
      />
    </svg>
  </button>
</div>
`;
